cdc/fifo.go

```diff
@@ -0,0 +1,260 @@
+package cdc
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+
+	"go.etcd.io/bbolt"
+)
+
+// bucketName is the name of the BoltDB bucket where queue items will be stored.
+var bucketName = []byte("fifo_queue")
+
+// ErrQueueEmpty is returned when First() is called on an empty queue.
+var ErrQueueEmpty = errors.New("queue is empty")
+
+// Queue is a persistent, disk-backed FIFO queue.
+// It is safe for concurrent use by multiple goroutines.
+type Queue struct {
+	db   *bbolt.DB
+	mu   sync.Mutex
+	cond *sync.Cond // cond is used to signal waiting Dequeue calls.
+}
+
+// NewQueue creates or opens a new persistent queue at the given file path.
+func NewQueue(path string) (*Queue, error) {
+	// Open the BoltDB database file at the given path.
+	// It will be created if it doesn't exist.
+	db, err := bbolt.Open(path, 0600, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open boltdb: %w", err)
+	}
+
+	// Start a read-write transaction to ensure our bucket exists.
+	err = db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(bucketName)
+		return err
+	})
+	if err != nil {
+		db.Close() // Close the db if bucket creation fails.
+		return nil, fmt.Errorf("failed to create bucket: %w", err)
+	}
+
+	q := &Queue{
+		db: db,
+	}
+	// Initialize the condition variable with the queue's mutex.
+	// This is crucial for coordinating the Dequeue and Enqueue operations.
+	q.cond = sync.NewCond(&q.mu)
+
+	return q, nil
+}
+
+// Close closes the underlying BoltDB database.
+func (q *Queue) Close() error {
+	return q.db.Close()
+}
+
+// Enqueue adds an item to the queue with a given index.
+// The FIFO order is determined by the index; lower indices are considered "earlier" in the queue.
+func (q *Queue) Enqueue(idx uint64, item []byte) error {
```

Comment:
Things to think about:

- only accept items which are not already present in the queue.
- only accept items which have a higher key than the last object already present in the queue (which will enforce the first point).

In other words while an attempt may be made to insert an object using an earlier or equal index to the last object in the queue, if an object is added with an index great than the last current object, that index must be exactly more more than the currently last index.

Reply:
Actually, not quite true. Some stuff goes through the Raft log, bumping the Raft Index, but would not result in a CDC event -- a strong query, for example.

Still, do not insert any older events. One an index is inserted, all future insertions must be a greater index.